core/type_inference/translate_jsonnet_to_lambda.py

```diff
@@ -0,0 +1,192 @@
+
+    elif isinstance(ast_, j_ast.LiteralNull):
+        return lam_ast.Identifier("null")
+
+    elif isinstance(ast_, j_ast.ObjectComprehension):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Self):
+        return lam_ast.Identifier("self")
+
+    elif isinstance(ast_, j_ast.SuperIndex):
+        return lam_ast.Identifier(ast_.index.value)
+
+    elif isinstance(ast_, j_ast.UnaryOp):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Var):
+        location = translate_location(ast_.location)
+        return lam_ast.Identifier(ast_.id, location)
+
+    else:
+        print(ast_.__class__)
+        raise Exception('Node is not found in jsonnet_ast\n')
+
+
+def build_record_type_constructor(fields):
+    var_type = {}
+    name_type = {}
+    for i, field in enumerate(fields):
+        name, _ = field
+        var = f'var{i}'
+        var_type[var] = TypeVariable()
+        name_type[name] = var_type[var]
+    record_type = TypeRowOperator(name_type)
+
+    def rec_build(i, n, var_type, record_type):
+        if n == 0:
+            return record_type
+        var = var_type[f'var{i}']
+        if (i == n-1):
+            return Function(var, record_type)
+        return Function(var, rec_build(i+1, n, var_type, record_type))
+
+    return rec_build(0, len(var_type), var_type, record_type)
+
+
+def build_letrec_and(field_keys, fields, body, env, location):
+    translated_fields = {}
+    for key in field_keys:
+        translated_body = translate_to_lambda_ast(fields[key], env)
+        name, loc = key
+        translated_fields[name] = (translated_body, loc)
+    return lam_ast.LetrecAnd(translated_fields, body, location)
+
+
+def get_next_record_id(env):
+    record_id = "record_{n}".format(n=env["__record_count__"])
+    env["__record_count__"] += 1
+    return record_id
+
+
+def get_next_plus_id(env):
+    plus_id = "plus_{n}".format(n=env["__plus_count__"])
+    env["__plus_count__"] += 1
+    return plus_id
+
+
+def apply_record(field_keys, record_id, env, location):
+    if not field_keys:
+        return lam_ast.Identifier(record_id)
+    name, loc = field_keys[-1]
+    return lam_ast.Apply(apply_record(field_keys[:-1], record_id, env, location),
+                         lam_ast.Identifier(name, loc),
+                         location)
+
+
+def translate_field_name(name):
+    location = translate_location(name.location)
+    if isinstance(name, j_ast.LiteralString):
+        return (name.value, location)
+    else:
+        raise Exception(f"Expected LiteralString but got {name.__class__}")
+
+
+def build_plus_op(left_arg, right_arg, env, location):
+    if isinstance(right_arg, j_ast.Object):
+        base_obj = translate_to_lambda_ast(left_arg, env)
+        child_obj = translate_to_lambda_ast(right_arg, env)
+        return lam_ast.Inherit(base_obj, child_obj, location)
+    else:
+        var = TypeVariable()
+        plus_id = get_next_plus_id(env)
```

Comment:
Do we really need different plusses for each use of "+"? I think having a single global "+" function of type `a -> a -> a` is sufficient.

Reply:
ahh, right.. we just make a copy of plus-type from the type environment. Yeah, one is enough.